pkg/controller/clusterautoscaler/clusterautoscaler_controller.go

```diff
@@ -17,43 +17,76 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
 const (
-	caImage          = "quay.io/bison/cluster-autoscaler:a554b4f5"
 	criticalPod      = "scheduler.alpha.kubernetes.io/critical-pod"
 	caServiceAccount = "cluster-autoscaler"
 )
 
 // Add creates a new ClusterAutoscaler Controller and adds it to the
 // Manager. The Manager will set fields on the Controller and Start it
 // when the Manager is Started.
-func Add(mgr manager.Manager) error {
-	return add(mgr, newReconciler(mgr))
+func Add(mgr manager.Manager, cfg *operator.Config) error {
+	r := newReconciler(mgr)
+
+	r.SetImage(cfg.ClusterAutoscalerImage)
+	r.SetReplicas(cfg.ClusterAutoscalerReplicas)
+	r.SetNamespace(cfg.ClusterAutoscalerNamespace)
+
+	return add(mgr, cfg, r)
 }
 
-// newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+// newReconciler returns a new Reconciler.
+func newReconciler(mgr manager.Manager) *Reconciler {
 	return &Reconciler{
 		client: mgr.GetClient(),
 		scheme: mgr.GetScheme(),
 	}
 }
 
+// processEvent is used in predicate functions.  It inspects metadata and
+// returns a boolean value representing whether the controller should process an
+// event for the object.
+func processEvent(meta metav1.Object, cfg *operator.Config) bool {
+	// Only process events for objects matching the configured resource name.
+	if meta.GetName() != cfg.ClusterAutoscalerName {
+		glog.Warningf("Not processing ClusterAutoscaler %s", meta.GetName())
+		return false
+	}
+
+	return true
+}
+
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func add(mgr manager.Manager, cfg *operator.Config, r reconcile.Reconciler) error {
 	// Create a new controller
 	c, err := controller.New("clusterautoscaler-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return err
 	}
 
+	// ClusterAutoscaler is effectively a singleton resource.  A
+	// deployment is only created if an instance is found matching the
+	// name set at runtime.
+	p := predicate.Funcs{
```

Comment:
don't we want to include delete?

Reply:
Delete doesn't do anything here. We rely on garbage collection and proper owner references to clean up resources. We could add it for completeness.

Reply:
Added delete and generic predicates here.